backend/dal/Services/Admin/Concrete/AgencyService.cs

```diff
@@ -100,35 +100,133 @@ public Agency Get(int id)
+        {
+            agency.ThrowIfNull(nameof(agency));
+
+            base.Add(agency);
+            this.Context.Entry(agency).State = EntityState.Detached;
+
+            // If the agency has been added as a sub-agency, then all users who are associated with the parent agency need to be updated.
+            if (agency.ParentId.HasValue)
+            {
+                var users = this.Context.Users.AsNoTracking().Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)).ToArray();
+                users.ForEach(u => agency.Users.Add(new UserAgency(u, agency)));
+            }
+        }
+
         /// <summary>
         /// Updates the specified agency in the datasource.
         /// </summary>
-        /// <param name="entity"></param>
-        /// <exception type="KeyNotFoundException">Entity does not exist in the datasource.</exception>
+        /// <param name="agency"></param>
+        /// <exception type="KeyNotFoundException">agency does not exist in the datasource.</exception>
         /// <returns></returns>
-        public override void Update(Agency entity)
+        public override void Update(Agency agency)
         {
-            entity.ThrowIfNull(nameof(entity));
+            agency.ThrowIfNull(nameof(agency));
+
+            var original = this.Context.Agencies.Find(agency.Id) ?? throw new KeyNotFoundException();
+            var updatedUsers = new List<User>();
+
+            // If the agency has become a sub-agency, or a parent-agency then users will need to be updated.
+            if (original.ParentId.HasValue && !agency.ParentId.HasValue)
+            {
+                // This agency has become a parent agency, all users associated with it will be removed
+                // because users should only be associated with a single parent.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
+
+                // Remove any user who are linked to the agency.
+                var users = this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(ua => ua.AgencyId == agency.Id));
+                users.ForEach(u => u.Agencies.Clear());
+                this.Context.UpdateRange(users);
+            }
+            else if (!original.ParentId.HasValue && agency.ParentId.HasValue)
+            {
+                // This agency has become a sub-agency, all original users need their list of agencies reduced to only this agency.
+                updatedUsers.AddRange(this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)));
+            }
+            else if (original.ParentId != agency.ParentId)
+            {
+                // Remove the sub-agency from curretly linked users and add it to the users who belong to the new parent agency.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
```

Comment:
I must be missing something, doesn't updatedUsers just add the agency being updated to all the users in the list? if so, how is this removing the sub-agency from users who are currently linked?

Reply:
It technically doesn't.  The DB DAL doesn't have access to the Keycloak DAL.
Users are only ever currently assigned to a single agency.  So the **User -> Agencies** relationship will currently only have the single agency they're assigned to.  However if they are assigned to a parent agency, they will **inherit** through logic all sub-agencies.  But the DB will only have the parent agency listed.
Most of the code here is simply returning all the users affected by the agency update.  In other words the DB doesn't need to be updated but keycloak does.  This is because in keycloak we keep a list of the parent agency and all sub-agencies the user has access to.
Now we return the list of users affected so that a separate routine can update keycloak.

I've noted in the comments that this isn't ideal, but we are dealing with two separate systems that need to be updated.